libclc/utils/libclc-remangler/LibclcRemangler.cpp

```diff
@@ -147,340 +123,606 @@ class BumpPointerAllocator {
-        OB << index;
-      OB << '_';
-      return true;
+private:
+  // Helper struct to aggregate information about types.
+  struct NodeKindInfo {
+    NodeKindInfo(Node::Kind K) : K(K) {}
+    NodeKindInfo(Node::Kind K, size_t NumElemsOrAS) : K(K) {
+      Data = NumElemsOrAS;
     }
-    return false;
-  }
-
-  void remangleOpenCLCName(const Node *nameNode, OutputBuffer &OB,
-                           bool Substitutable, bool isNameRoot = true) {
-    if (Substitutable && remangleSub(nameNode, OB))
-      return;
-    switch (nameNode->getKind()) {
-    case Node::Kind::KNameType: {
-      const NameType *name = static_cast<const NameType *>(nameNode);
-      OB << name->getName().size();
-      OB << name->getName();
-      break;
+    NodeKindInfo(Node::Kind K, itanium_demangle::Qualifiers Quals) : K(K) {
+      Data = 0;
+      if (Quals & itanium_demangle::Qualifiers::QualConst)
+        Data |= clang::Qualifiers::TQ::Const;
+      if (Quals & itanium_demangle::Qualifiers::QualVolatile)
+        Data |= clang::Qualifiers::TQ::Volatile;
+      if (Quals & itanium_demangle::Qualifiers::QualRestrict)
+        Data |= clang::Qualifiers::TQ::Restrict;
     }
-    case Node::Kind::KNestedName: {
-      if (isNameRoot)
-        OB << 'N';
-      const NestedName *nestedName = static_cast<const NestedName *>(nameNode);
-      remangleOpenCLCName(nestedName->Qual, OB, Substitutable,
-                          /* isNameRoot= */ false);
-      remangleOpenCLCName(nestedName->Name, OB, /* Substitutable= */ false,
-                          /* isNameRoot= */ false);
-      if (isNameRoot)
-        OB << 'E';
-      break;
-    }
-    case Node::Kind::KNameWithTemplateArgs: {
-      const NameWithTemplateArgs *templateName =
-          static_cast<const NameWithTemplateArgs *>(nameNode);
-      assert(templateName->TemplateArgs->getKind() ==
-             Node::Kind::KTemplateArgs);
-      remangleOpenCLCName(templateName->Name, OB, /* Substitutable= */ false,
-                          /* isNameRoot= */ false);
-      OB << 'I';
-      const TemplateArgs *templateArgs =
-          static_cast<const TemplateArgs *>(templateName->TemplateArgs);
-      for (auto templateArgType : templateArgs->getParams())
-        remangleOpenCLCType(templateArgType, OB);
-      OB << 'E';
-      break;
+    NodeKindInfo(Node::Kind K, const char *NestedName, size_t NestedNameSize)
+        : K(K) {
+      Data = 0;
+      DataStr.assign(NestedName, NestedNameSize);
     }
-    default: {
-      OutputBuffer errorTypeOut;
-      errorTypeOut << "Unhandled name : ";
-      nameNode->print(errorTypeOut);
-      errorTypeOut << "\n";
-      errs() << errorTypeOut.getBuffer();
-      free(errorTypeOut.getBuffer());
-      Failed = true;
+    Node::Kind K;
+    size_t Data = 0;
+    std::string DataStr{};
+  };
+
+private:
+  // Construct FunctionDecl from return, argument and template types.
+  FunctionDecl *createKernelDecl(clang::QualType &RetTy,

```

Comment:
Should all of these be non-const references? It looks like the only caller doesn't ever inspect these as output values.

I suspect you want the vectors to be const references and the qual type to be passed as a value type.

Reply:
You're right, the only slight complication is the `ArgTy` as the function modifies it as it goes. I could turn it into a value type and move it in place, or leave as non-const ref?

Reply:
The only caller of this function doesn't inspect `ArgTys` after the call returns, so I'd make it a const ref and pay for the copy operation (it should be cheap, even hundreds of args would not be that expensive). That also makes it more clear within the function as to whether you're looking at the original argument types or the adjusted ones. WDYT?